### PR TITLE
fix Issue 22321 - ImportC: non-static arrays can.t be initialized by …

### PR DESCRIPTION
**Pull request by WalterBright**
…an initializer list

Fix it by converting a C initializer to an ArrayInitializer, because D can handle it.

### CODE REVIEW

File: src/dmd/initsem.d [hunk 1]
```diff
@@ -363,10 +363,10 @@ extern(C++) Initializer initializerSemantic(Initializer init, Scope* sc, ref Typ
             if (length > i.dim)
                 i.dim = length;
         }
-        if (t.ty == Tsarray)
+        if (auto tsa = t.isTypeSArray())
         {
-            uinteger_t edim = (cast(TypeSArray)t).dim.toInteger();
-            if (i.dim > edim)
+            uinteger_t edim = tsa.dim.toInteger();
+            if (i.dim > edim && !(tsa.isIncomplete() && (sc.flags & SCOPE.Cfile)))
```
Comment (RazvanN7) [inline src/dmd/initsem.d:369]:
```suggestion
            if (i.dim > edim && !((sc.flags & SCOPE.Cfile) && tsa.isIncomplete()))
```

File: src/dmd/initsem.d [hunk 2]
```diff
@@ -721,17 +721,17 @@ extern(C++) Initializer initializerSemantic(Initializer init, Scope* sc, ref Typ
         {
             //printf(" type %s i %d dim %d dil.length = %d\n", t.toChars(), cast(int)i, cast(int)dim, cast(int)dil.length);
             auto tn = t.nextOf().toBasetype();
-            if (auto tna = tn.isTypeDArray())
+            auto tnsa = tn.isTypeSArray();
+            if (tnsa && tnsa.isIncomplete())
```
Comment (RazvanN7) [inline src/dmd/initsem.d:725]: Shouldn't this condition also check that that we are in cfile scope?

**Comment by RazvanN7** [comment]
Also, rebasing will fix the azure failures.

File: src/dmd/initsem.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/initsem.d:725]: That's implicit in it being in CInitializer code.

File: src/dmd/initsem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/initsem.d:369]: Rationale?

**Comment by WalterBright** [comment]
I did a rebase:
```
mercury~/forks/dmd fix22312> git rebase stable
First, rewinding head to replay your work on top of it...
mercury~/forks/dmd fix22312> git push -f origin fix22312
Total 0 (delta 0), reused 0 (delta 0)
To git@github.com:WalterBright/dmd.git
 + a28277c...f7f8456 fix22312 -> fix22312 (forced update)
mercury~/forks/dmd fix22312>
```
but github didn't restart the tests.

File: src/dmd/initsem.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/initsem.d:369]: This code will be executed for both D and C code, however, the former should be the common case, in which case `tsa.isIncomplete` will be evaluated only if necessary.

File: src/dmd/initsem.d (hunk 2 above)
Comment (RazvanN7) [reply src/dmd/initsem.d:725]: Oh, I missed the fact that we are in `visitC`.

**Review by ibuclaw — APPROVED** [review]

File: src/dmd/initsem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/initsem.d:369]: `isComplete` is a trivial function.

**Pull request merged**